gobblin-yarn/src/main/java/gobblin/yarn/YarnService.java

```diff
@@ -340,6 +369,60 @@ private String buildContainerCommand(Container container) {
   }
 
   /**
+   * Check the exit status of a completed container and see if the replacement container
+   * should try to be started on the same node. Some exit status indicates a disk or
+   * node failure and in such cases the replacement container should try to be started on
+   * a different node.
+   */
+  private boolean shouldStickToTheSameNode(int containerExitStatus) {
+    switch (containerExitStatus) {
+      case ContainerExitStatus.DISKS_FAILED:
+      case ContainerExitStatus.ABORTED:
+        // Mostly likely this exit status is due to node failures because the
+        // application itself will not release containers.
+        return false;
+      default:
+        // Stick to the same node for other cases if host affinity is enabled.
+        return this.containerHostAffinityEnabled;
+    }
+  }
+
+  /**
+   * Handle the completion of a container. A new container will be requested to replace the one
+   * that just exited. Depending on the exit status and if container host affinity is enabled,
+   * the new container may or may not try to be started on the same node.
+   */
+  private void handleContainerCompletion(ContainerStatus containerStatus) {
```

Comment:
My understanding is that since we have long running containers, the only time a container should complete is if there was either (1) on error in the container that caused it to crash, or (2) we are trying to shutdown the instance. If my understanding is correct, can you add that to the javadocs of this method. It's not terribly clear to somewhat not super familiar with the design.

Does `AMRMClientCallbackHandler.onContainersCompleted` get invoked when we try to shutdown the whole instance? Do we need any additional handling to ensure we have a clean shutdown.